torch/cuda/graphs.py

```diff
@@ -0,0 +1,89 @@
+import ctypes
+import gc
+import torch
+
+from ._utils import _dummy_type
+
+
+if not hasattr(torch._C, '_CudaStreamBase'):
+    # Define dummy base classes
+    torch._C.__dict__['CUDAGraph'] = _dummy_type('CUDAGraph')
+    torch._C.__dict__['graph_pool_handle'] = _dummy_type('graph_pool_handle')
+
+from torch._C import (
+    CUDAGraph,
+    graph_pool_handle
+)
+
+
+class graph(object):
+    r"""Context-manager that captures CUDA work into a :class:`torch.cuda.CUDAGraph`
+    object for later replay.
+
+    See :ref:`CUDA Graphs <cuda-graph-semantics>` for a general introduction,
+    detailed use, and constraints.
+
+    Arguments:
+        graph (torch.cuda.CUDAGraph): Graph object used for capture.
+        pool: Opaque token (returned by a call to :func:`~torch.cuda.graph_pool_handle()` or :meth:`other_Graph_instance.pool()<torch.cuda.CUDAGraph.pool>`) hinting this graph's capture may share memory from the specified pool.
+        stream (torch.cuda.Stream, optional): If supplied, will be set as the current ambient stream in the context.  If not supplied, ``Graph`` sets its own internal side stream as the ambient stream in the context.
+
+    .. note::
+        For effective memory sharing, if you pass a ``pool`` used by a previous capture and the previous capture
+        used an explicit ``stream`` argument, you should pass the same ``stream`` argument to this capture.
```

Comment:
What happens if you share a pool, but don't send a stream to one or both `graph` context managers?

Reply:
The underlying point is that memory is siloed by pool and stream. If the new capture uses the same pool as a previous capture, but a different stream, it won't actually reuse any of the previous capture's allocations (they'll come from the same PrivatePool, but a different stream silo, which defeats the purpose of using the same Pool).

Man, I wish i had integrated cudaMallocAsync first, then no one would have to worry about pools. But cudaMallocAsync wasnt graphable til 11.4 and mlperf needed graphs months earlier.